### PR TITLE
Convert ClientSession to follow builder pattern

### PR DESCRIPTION
**Pull request by akshayrai**
## Description

Refactor ClientSession to follow builder pattern consistently

> Is this change a fix, improvement, new feature, refactoring, or other?

refactoring

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

client library

> How would you describe this change to a non-technical end user or system administrator?

N/A

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

### CODE REVIEW

File: client/trino-client/src/main/java/io/trino/client/ClientSession.java [hunk 1]
```diff
@@ -315,21 +322,81 @@ private Builder(ClientSession clientSession)
...
+        public Builder withClientInfo(String clientInfo)
+        {
+            this.clientInfo = clientInfo;
+            return this;
+        }
+
         public Builder withCatalog(String catalog)
         {
-            this.catalog = requireNonNull(catalog, "catalog is null");
+            this.catalog = catalog;
```
[the remark this answers is not in the document]
Comment (akshayrai) [reply client/trino-client/src/main/java/io/trino/client/ClientSession.java:369]: I too feel keeping all the checks in the constructor (one place) should be enough and we can remove the RNN checks from the methods. Will update accordingly.

File: client/trino-client/src/main/java/io/trino/client/ClientSession.java [hunk 2]
```diff
@@ -66,7 +71,7 @@ public static ClientSession stripTransactionId(ClientSession session)
                 .build();
     }
 
-    public ClientSession(
+    private ClientSession(
```
[the remark this answers is not in the document]
Comment (akshayrai) [reply client/trino-client/src/main/java/io/trino/client/ClientSession.java:74]: If you take a look at [ClientOptions](https://github.com/trinodb/trino/blob/master/client/trino-cli/src/main/java/io/trino/cli/ClientOptions.java#L105), the user and sessionUser optional references are not initialized (null by default). There are many other optional variables in ClientOptions which are also not initialized. I do not know if that is intentional. I don't mind doing it just for user and sessionUser/principal (i.e., set to Optional.empty) if that makes sense from the context of this PR?

File: client/trino-client/src/main/java/io/trino/client/ClientSession.java [hunk 3]
```diff
@@ -273,23 +278,25 @@ public String toString()
         private Optional<String> principal;
         private Optional<String> user;
         private String source;
-        private Optional<String> traceToken;
-        private Set<String> clientTags;
+        private Optional<String> traceToken = Optional.empty();
```
[the remark this answers is not in the document]
Comment (akshayrai) [reply client/trino-client/src/main/java/io/trino/client/ClientSession.java:281]: will address this based on how the previous comment concludes.

File: client/trino-client/src/main/java/io/trino/client/ClientSession.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (phd3) [reply client/trino-client/src/main/java/io/trino/client/ClientSession.java:74]: I think some items can be `null` there, but I'm mainly talking about the `Optional<>` fields. Having them as null is confusing. 

Looks like Picocli should take care of them as far as ClientOptions is concerned: https://picocli.info/#_optionalt So I don't think any backwards incompatible change would happen here. cc @electrum

[the remark this answers is not in the document]
Comment (akshayrai) [reply client/trino-client/src/main/java/io/trino/client/ClientSession.java:74]: @phd3 cool, that makes sense. thanks for that pointed to Picocli. I will update accordingly.

File: testing/trino-tests/src/test/java/io/trino/execution/TestFinalQueryInfo.java [hunk 4]
```diff
@@ -63,27 +61,17 @@ private static QueryId startQuery(String sql, DistributedQueryRunner queryRunner
...
-                    ImmutableMap.of(),
-                    ImmutableMap.of(),
-                    ImmutableMap.of(),
-                    null,
-                    new Duration(2, MINUTES),
-                    true);
+            ClientSession clientSession = ClientSession.builder()
+                    .withServer(queryRunner.getCoordinator().getBaseUrl())
+                    .withPrincipal(Optional.of("user"))
+                    .withUser(Optional.empty())
```
Comment (phd3) [inline testing/trino-tests/src/test/java/io/trino/execution/TestFinalQueryInfo.java:67]: redundant

File: testing/trino-testing/src/main/java/io/trino/testing/AbstractTestingTrinoClient.java [hunk 5]
```diff
@@ -147,27 +147,28 @@ private static ClientSession toClientSession(Session session, URI server, Durati
...
-                session.getPreparedStatements(),
-                getRoles(session),
-                session.getIdentity().getExtraCredentials(),
-                session.getTransactionId().map(Object::toString).orElse(null),
-                clientRequestTimeout,
-                true);
+        return ClientSession.builder()
+                .withServer(server)
+                .withPrincipal(Optional.of(session.getIdentity().getUser()))
+                .withUser(Optional.empty())
```
Comment (phd3) [inline testing/trino-testing/src/main/java/io/trino/testing/AbstractTestingTrinoClient.java:153]: redundant

**Review by phd3 — APPROVED** [review]
LGTM % couple comments. Please squash

File: testing/trino-tests/src/test/java/io/trino/execution/TestFinalQueryInfo.java (hunk 4 above)
Comment (akshayrai) [reply testing/trino-tests/src/test/java/io/trino/execution/TestFinalQueryInfo.java]: removed

File: testing/trino-testing/src/main/java/io/trino/testing/AbstractTestingTrinoClient.java (hunk 5 above)
Comment (akshayrai) [reply testing/trino-testing/src/main/java/io/trino/testing/AbstractTestingTrinoClient.java]: removed

File: client/trino-client/src/main/java/io/trino/client/ClientSession.java [hunk 6]
```diff
@@ -315,27 +322,87 @@ private Builder(ClientSession clientSession)
             compressionDisabled = clientSession.isCompressionDisabled();
         }
 
+        public Builder withServer(URI server)
```
Comment (martint) [inline client/trino-client/src/main/java/io/trino/client/ClientSession.java:325]: Remove the `with` prefix from all these methods.

Comment (akshayrai) [reply client/trino-client/src/main/java/io/trino/client/ClientSession.java]: ack! removed all the "with" prefix.

**Comment by phd3** [comment]
Thanks @akshayrai can you please squash?

**Comment by phd3** [comment]
CI hit https://github.com/trinodb/trino/issues/13107

**Comment by phd3** [comment]
CI hit https://github.com/trinodb/trino/issues/13633

**Comment by phd3** [comment]
Merged, thanks!

**Pull request merged**